core/src/serve_repair.rs

```diff
@@ -258,6 +258,62 @@ pub enum RepairProtocol {
     },
 }
 
+const REPAIR_REQUEST_PONG_SERIALIZED_BYTES: usize = PUBKEY_BYTES + HASH_BYTES + SIGNATURE_BYTES;
+const REPAIR_REQUEST_HEADER_SERIALIZED_BYTES: usize =
+    SIGNATURE_BYTES + PUBKEY_BYTES + PUBKEY_BYTES + 8 + 4;
+
+fn check_well_formed_repair_request(packet: &Packet, stats: &mut ServeRepairStats) -> Result<()> {
+    if let Some(discriminator) = packet
+        .data(..4)
+        .and_then(|data| <[u8; 4]>::try_from(data).ok())
+        .map(u32::from_le_bytes)
+    {
+        let expected_size = match discriminator {
+            0..=6 => {
+                // deprecated requests
+                stats.err_unsigned += 1;
+                return Err(Error::from(RepairVerifyError::Unsigned));
+            }
+            7 => 4 + REPAIR_REQUEST_PONG_SERIALIZED_BYTES, // Pong
+            8 => 4 + REPAIR_REQUEST_HEADER_SERIALIZED_BYTES + 8 + 8, // WindowIndex
+            9 => 4 + REPAIR_REQUEST_HEADER_SERIALIZED_BYTES + 8 + 8, // HighestWindowIndex
+            10 => 4 + REPAIR_REQUEST_HEADER_SERIALIZED_BYTES + 8, // Orphan
+            11 => 4 + REPAIR_REQUEST_HEADER_SERIALIZED_BYTES + 8, // AncestorHashes
+            _ => {
+                // invalid repair request type
+                stats.err_malformed += 1;
+                return Err(Error::from(RepairVerifyError::Malformed));
+            }
+        };
+        if packet.meta().size == expected_size {
+            return Ok(());
+        }
+    }
+    stats.err_malformed += 1;
+    Err(Error::from(RepairVerifyError::Malformed))
+}
+
+fn discard_malformed_repair_requests(
+    batch: &mut PacketBatch,
+    stats: &mut ServeRepairStats,
+    cluster_type: ClusterType,
+) -> usize {
+    let mut retained = 0;
+    for packet in batch.iter_mut() {
+        match check_well_formed_repair_request(packet, stats) {
+            Ok(()) => retained += 1,
+            Err(Error::RepairVerify(RepairVerifyError::Unsigned)) => match cluster_type {
+                ClusterType::Testnet | ClusterType::Development => {
```

Comment:
We expect repair requests to be signed on testnet and development clusters but not on MNB or Devnet? Can you help me understand?

Reply:
discussed offline